README.md
# Openfin react starter `concise version`

[![version][version-badge]][CHANGELOG] [![license][license-badge]][LICENSE]
[![Build Status](https://travis-ci.com/openfin-js-app/openfin-react-concise.svg?branch=master)](https://travis-ci.com/openfin-js-app/openfin-react-concise)
[![Coverage Status](https://coveralls.io/repos/github/openfin-js-app/openfin-react-concise/badge.svg?branch=master)](https://coveralls.io/github/openfin-js-app/openfin-react-concise?branch=master)

A concise version of openfin react starter

![](https://albertleigh.github.io/openfin-react-latest/img/screenshoot.gif)

* Design with simplicity in mind
* Based on create-react-app@3.0.1
* Integrate Jest test framework
* Integrate Material-UI@4.0.2
* Support basic frameless window features: move, minimize, maximize, close and resize
* Support general sidebar
* Support general primary/success/error/info/warning snackbar
* Support spawn a child window
* Version tag injected via dotenv
* Support general app config
* Support IndexDB based client side config
* Support snap&dock out of box
* Support Notification Window
* Support HOC Config context
* Embrace react hook apis

[LICENSE]: ./LICENSE.md
[CHANGELOG]: ./CHANGELOG.md

[version-badge]: https://img.shields.io/badge/version-1.1.0-green.svg
[license-badge]: https://img.shields.io/badge/license-MIT-green.svg




CHANGELOG.md
## [1.1.0] 2019-06-19
- Support PUBLIC_URL
- Restore DockerFile

## [1.0.1] 2019-06-16
- Update dependencies to
    - @material-ui/core @ ^4.0.2
    - create-react-app @ 3.0.1
- Remove dependencies
    - material-ui-pickers @ ^2.2.4

## [0.40.20] 2019-04-23
- First initial stable release

scripts/server.js
#!/usr/bin/env node

process.env.REACT_APP_ENV = 'production';
process.env.BABEL_ENV = 'production';
process.env.NODE_ENV = 'production';

require('../config/env');

const express = require('express');
const logger = require('morgan');
const http = require('http');
const paths = require('../config/paths');
const debug = require('debug')('openfin-react-concise:scripts/server');

const configApis = require('./routers/config');

const {
    normalizePort,onError,
} = require('./utils/serverUtils');

const app = express();

app.use(logger('dev'));
app.use(configApis);
app.use(express.static(paths.appBuild));

app.use((req,res,next)=>{
    if (req.accepts('html')){
        res.status(200);
        res.sendFile(paths.appBuild+'/index.html');
    }else{
        res.status(404);
        if(req.accepts('json')){
            res.send({error:'Not found'});
        }
        res.type('txt'),send('Not found');
    }
});

var port = normalizePort(process.env.PORT || '3000');
app.set('port',port);

var server = http.createServer(app);

server.listen(port);
server.on('error',onError);
server.on('listening',onListening);

function onListening() {
    var addr = server.address();
    var bind = typeof addr === 'string'
        ? 'pipe ' + addr
        : 'port ' + addr.port;
    debug('Listening on ' + bind);
    console.log('Listening on ' + bind);
}

src/routes/Dashboard.ts
import AccessibilityIcon from '@material-ui/icons/Accessibility';
import ReportIcon from '@material-ui/icons/Report';
import SettingsIcon from '@material-ui/icons/Settings';

import { RouteItem, ReportView } from 'react-openfin-mat-impl';

import { MyConfigView } from '../constants/configTabs';

import Accessibility from '../views/Accessibility/Accessibility';

import { mapToPublicPathname } from './utils';

const dashboardRoutes:RouteItem[]=[
    {
        path:'/dashboard/accessibility',
        sidebarName:"accessibility.sidebar",
        navbarName:"accessibility.navbar",
        icon: AccessibilityIcon,
        component: Accessibility,
    },
    {
        path:'/dashboard/config',
        sidebarName:"configuration.sidebar",
        navbarName:"configuration.navbar",
        icon: SettingsIcon,
        component: MyConfigView,
    },
    {
        path:'/dashboard/report',
        sidebarName:"report.sidebar",
        navbarName:"report.navbar",
        icon: ReportIcon,
        component: ReportView,
    },
];

export default dashboardRoutes.map(mapToPublicPathname);

scripts/server.openfin.js
// Do this as the first thing so that any code reading it knows the right env.
process.env.REACT_APP_ENV = 'production';
process.env.BABEL_ENV = 'production';
process.env.NODE_ENV = 'production';

const paths = require('../config/paths');
const spawn = require('child_process').spawn;
const chalk = require('chalk');
const log = console.log;
const tcpPortUsed = require('tcp-port-used');

const { connect } = require('hadouken-js-adapter');

require('../config/env');
const DEFAULT_PORT = parseInt(process.env.PORT, 10) || 3000;
let expressServer = null;

async function startServer() {
    const env = Object.create(process.env);
    expressServer = spawn( 'node',[paths.appScript+'/server.js'], env );
    expressServer.stdout.on('data',(data)=>{
        log(chalk.cyan(Buffer.from(data,'binary').toString()));
    });
    expressServer.stderr.on('data',(data)=>{
        try{
            log(chalk.red(Buffer.from(data,'binary').toString()));
        }catch(e){
            log(chalk.red(data));

        }
    });
    expressServer.on('close',(data)=>{
        log(chalk.yellow(Buffer.from(data,'binary').toString()));
    });
}

async function launchApp(){
    const fin = await connect({
        uuid:'openfin_react_concise',
        runtime:{
            version: process.env.HADOUKEN_VERSION,
        }
    });
    const version = await fin.System.getVersion();
    log(chalk.green("Connected to Hadouken version", version));

    const app = await fin.Application.create({
        "name":"Openfin Concise",
        "url":`http://localhost:${DEFAULT_PORT}/index.html`,
        "uuid":process.env.REACT_APP_FIN_UUID,
        "applicationIcon":`http://localhost:${DEFAULT_PORT}/favicon.ico`,
        "autoShow":false,
        "saveWindowsSate":false,
        "resizable":true,
        "frame":false,
        "defaultTop":parseInt(process.env.REACT_APP_NEW_WINDOW_TOP,10),
        "defaultLeft":parseInt(process.env.REACT_APP_NEW_WINDOW_LEFT,10),
        "defaultWidth":728,
        "defaultHeight":450,
        "minWidth":420,
        "minHeight":300
    });

    log(chalk.green(`connecting to http://localhost:${DEFAULT_PORT}`));

    app.addListener('closed',()=>{
        if (expressServer){
            expressServer.kill('SIGINT');
        }
        process.exit(0);
    });

    await app.run();
}

startServer();

tcpPortUsed.waitUntilUsed(DEFAULT_PORT,1000,240000)
    .then(
        ()=>{
            log(chalk.green('starting openfin'));
            launchApp().then(() => {
                log(chalk.green('starting openfin success'));
                console.log("success");
            }).catch((err) => {
                log(chalk.red("Error trying to connect,", err.message));
                log(chalk.red(err.stack));
            });
        },
        (err)=>{
            log(chalk.red(err));
        }
    );

scripts/start.openfin.js
// Do this as the first thing so that any code reading it knows the right env.
process.env.REACT_APP_ENV = 'development';
process.env.BABEL_ENV = 'development';
process.env.NODE_ENV = 'development';
const os =require('os');
const paths = require('../config/paths');
const spawn = require('child_process').spawn;
const chalk = require('chalk');
const log = console.log;
const tcpPortUsed = require('tcp-port-used');

const { connect } = require('hadouken-js-adapter');

require('../config/env');
const DEFAULT_PORT = parseInt(process.env.PORT, 10) || 3000;
let webDevServer = null;

async function startWebpackDevServer() {
    let cmd = 'npm';
    if (os.platform() === 'win32'){
        cmd = 'npm.cmd';
    }
    webDevServer = spawn(cmd,['run','start']);
    webDevServer.stdout.on('data',(data)=>{
        log(chalk.cyan(Buffer.from(data,'binary').toString()));
    });
    webDevServer.stderr.on('data',(data)=>{
        log(chalk.magenta(Buffer.from(data,'binary').toString()));
    });
    webDevServer.on('close',(data)=>{
        log(chalk.yellow(Buffer.from(data,'binary').toString()));
    });
}

async function launchApp(){
    const fin = await connect({
        uuid:'openfin_react_concise',
        runtime:{
            version: process.env.HADOUKEN_VERSION,
            arguments:"--v=1 --remote-debugging-port=9090 --enable-crash-reporting"
        }
    });
    const version = await fin.System.getVersion();
    log(chalk.green("Connected to Hadouken version", version));

    const app = await fin.Application.create({
        "name":"Openfin Concise [dev]",
        "url":`http://localhost:${DEFAULT_PORT}/index.html`,
        "uuid":process.env.REACT_APP_FIN_UUID,
        "applicationIcon":`http://localhost:${DEFAULT_PORT}/favicon.ico`,
        "autoShow":false,
        "saveWindowsSate":false,
        "resizable":true,
        "frame":false,
        "defaultTop":parseInt(process.env.REACT_APP_NEW_WINDOW_TOP,10),
        "defaultLeft":parseInt(process.env.REACT_APP_NEW_WINDOW_LEFT,10),
        "defaultWidth":728,
        "defaultHeight":450,
        "minWidth":420,
        "minHeight":300
    });

    log(chalk.green(`connecting to http://localhost:${DEFAULT_PORT}`));

    app.addListener('closed',()=>{
        if (webDevServer){
            webDevServer.kill('SIGINT');
        }
        process.exit(0);
    });

    await app.run();
}


startWebpackDevServer();

tcpPortUsed.waitUntilUsed(DEFAULT_PORT,1000,240000)
    .then(
        ()=>{
            log(chalk.green('starting openfin'));
            launchApp().then(() => {
                log(chalk.green('starting openfin success'));
                console.log("success");
            }).catch((err) => {
                log(chalk.red("Error trying to connect,", err.message));
                log(chalk.red(err.stack));
            });

        },
        (err)=>{
            log(chalk.red(err));
        }
    );